### PR TITLE
Add ConfigEnum class

### PR DESCRIPTION
**Pull request by Robbybp**
## Summary/Motivation:
Sometimes would like to accept user input in form of an Enum, int, or string, each of which is translated to the corresponding Enum value.

## Changes proposed in this PR:
- Add ConfigEnum class, which implements the from_enum_or_string method, intended for use as a domain in a ConfigValue
- Test

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/common/config.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 import os
 import platform
 
+import enum
```
Comment (jsiirola) [inline pyomo/common/config.py:14]: FYI: This places a hard dependency on `enum34` for Python 2.7.  I am not entirely sure how to resolve this.

Comment (Robbybp) [reply pyomo/common/config.py:14]: This is an unfamiliar problem for me. The issue is that `enum` was added in Python 3.4 and backported to Python 2.7, but there is no guarantee that a user on 2.7 will have the backported enum module? 

My first thought would be to put the import and class definition in a try-except, then mark the test with skipIf.

Comment (jsiirola) [reply pyomo/common/config.py:14]: The problem is that this isn't in a test: you need `enum` in the *definition* of the `ConfigEnum` class, so the entire `ConfigEnum` class wouldn't be available if enum wasn't present (only an issue for Python 2.7 at this point).  I think the right answer is to add enum34 to the dependencies for python2.7 in setup.py, but adding hard dependencies is something that we should at least make sure more developer weigh in on.

I will update this PR to include the new dependency and then wait for others to weigh in.

Comment (qtothec) [reply pyomo/common/config.py:14]: @jsiirola It helps that you are a main proponent of keeping 2.7 support in the first place. Fwiw, I believe it makes sense to keep support as long as it doesn't make life too difficult. Most people are content if you don't mess with 3.x.

**Comment by michaelbynum** [comment]
I think a dependency on enum34 is perfectly reasonable.

**Pull request merged**